[libc][math][c23] Add dfma{l,f128} and dsub{l,f128} C23 math functions

aaryanshukla opened the pull request:
- **[libc][math]: implemented dfma and dsub**
- **functions**
- **fixed entrypoints**
- **adding**
- **everything but index.rst**
- **[libc][math][c23]**

aaryanshukla committed:
[libc][math]: implemented dfma and dsub

aaryanshukla committed:
functions

aaryanshukla committed:
fixed entrypoints

aaryanshukla committed:
adding

aaryanshukla committed:
everything but index.rst

aaryanshukla committed:
[libc][math][c23]

aaryanshukla committed:
formatting and spacing

overmighty committed:
Update libc/test/src/math/smoke/SubTest.h dependencies

overmighty reviewed (commented):
I would suggest you use https://github.com/overmighty/llvm-dev-utils/blob/master/llvm_libc_add_math_function.py to save time. For example:

```sh
for type in 'l' 'f128'; do
  ~/projects/llvm-dev-utils/llvm_libc_add_math_function.py "${type}" 'double' 'dfma' 'TYPE x, TYPE y, TYPE z' 'FmaTest' 'LIST_NARROWING_FMA_TESTS'
  ~/projects/llvm-dev-utils/llvm_libc_add_math_function.py "${type}" 'double' 'dsub' 'TYPE x, TYPE y' 'SubTest' 'LIST_SUB_TESTS'
done
```

Right now it generates the header, implementation file boilerplate, and smoke test based on a test template header. It doesn't generate MPFR tests yet.

overmighty commented:
Since you're adding smoke tests that use SubTest.h, I pushed a commit that updates the SubTest.h dependencies, if you don't mind.

overmighty commented on libc/test/src/math/CMakeLists.txt:
```diff
@@ -2368,6 +2368,58 @@ add_fp_unittest(
+    dfmaf128_test.cpp
+  HDRS
+    FmaTest.h
+  DEPENDS
+    libc.src.math.dfmaf128
```
Actually, we don't support `float128` in MPFRWrapper/MPFRUtils, so this test should be removed.

overmighty commented on libc/test/src/math/CMakeLists.txt:
```diff
@@ -2368,6 +2368,58 @@ add_fp_unittest(
+    libc.src.math.dsubl
+)
+
+add_fp_unittest(
+  dsubf128_test
+  NEED_MPFR
+  SUITE
+    libc-math-unittests
+  SRCS
+    dsubf128_test.cpp
+  HDRS
+    SubTest.h
+  DEPENDS
+    libc.src.math.dsubf128
+)
```
This MPFR test should be removed since it uses `float128`.

overmighty commented:
You should test locally with either:

```
cmake --build path/to/your/build/directory --target libc-math-unittests
```

or:

```
cmake --build path/to/your/build/directory --target check-libc
```

Part of the reason why the CI failed is that you need to add new explicit template specializations in libc/utils/MPFRWrapper/MPFRUtils.cpp for functions that take `long double` inputs and output a `double`.

aaryanshukla committed:
addressed comments

included long double inp and double output template
removed from riscv entrypoints
removed dfmaf128 and dsubf128 from MPFR

overmighty commented on libc/config/baremetal/arm/entrypoints.txt:
```diff
@@ -247,6 +247,8 @@ set(TARGET_LIBM_ENTRYPOINTS
     libc.src.math.cos
     libc.src.math.cosf
     libc.src.math.coshf
+    libc.src.math.dfmal
+    libc.src.math.dsubl
     libc.src.math.erff
```
I think you should remove the new entrypoints from all baremetal targets, not just baremetal RISC-V.

aaryanshukla commented:
@overmighty I see that the build failed not due to my changes. Am I still good to merge?

overmighty commented on libc/src/math/generic/dfmaf128.cpp:
```diff
@@ -0,0 +1,25 @@
+//===-- Implementation of dfmaf128 header -----------------------*- C++ -*-===//
```
```suggestion
//===-- Implementation of dfmaf128 function -------------------------------===//
```

You can just copy-paste it, or even just click the "Commit suggestion" button on GitHub.

overmighty commented on libc/src/math/generic/dfmaf128.cpp:
```diff
@@ -0,0 +1,25 @@
+#ifndef LLVM_LIBC_SRC_MATH_DFMAf128_H
+#define LLVM_LIBC_SRC_MATH_DFMAf128_H
+
+#include "src/math/dfmaf128.h"
+#include "src/__support/FPUtil/FMA.h"
+#include "src/__support/macros/config.h"
+#include "src/__support/macros/properties/types.h"
```
```suggestion
#include "src/math/dfmaf128.h"
#include "src/__support/FPUtil/FMA.h"
#include "src/__support/common.h"
#include "src/__support/macros/config.h"
```

aaryanshukla replied:
fixed all

aaryanshukla committed:
fixed clang formatting error

aaryanshukla committed:
Update libc/src/math/generic/dfmaf128.cpp

Co-authored-by: OverMighty <its.overmighty@gmail.com>

aaryanshukla committed:
addressed comments

removed entrypoints from riscv

aaryanshukla commented:
@overmighty @lntue should I merge this now?

overmighty commented on libc/test/src/math/smoke/CMakeLists.txt:
```diff
@@ -358,6 +358,62 @@ add_fp_unittest(
+    dsubl_test.cpp
+  HDRS
+    SubTest.h
+  DEPENDS
+    libc.src.errno.errno
+    libc.hdr.fenv_macros
+    libc.src.math.dsubl
+)
+
+add_fp_unittest(
+  dsubf128_test
+  SUITE
+    libc-math-smoke-tests
+  SRCS
+    dsubf128_test.cpp
+  HDRS
+    SubTest.h
+  DEPENDS
+    libc.src.errno.errno
+    libc.hdr.fenv_macros
```
Should have been `libc.hdr.errno_macros` instead of `libc.src.errno.errno` here.

overmighty commented on libc/test/src/math/CMakeLists.txt:
```diff
@@ -2368,6 +2368,37 @@ add_fp_unittest(
+    dsubl_test.cpp
+  HDRS
+    SubTest.h
+  DEPENDS 
+    libc.src.errno.errno
+    libc.hdr.fenv_macros
+    libc.src.math.dsubl
```
libc/test/src/math/SubTest.h doesn't use `errno` or `fenv_macros`: https://github.com/llvm/llvm-project/blob/ff8a27f7393d0441aa220d0ecb3e1c7485c4449b/libc/test/src/math/SubTest.h#L12-L15

```suggestion
  DEPENDS
    libc.src.math.dsubl
```

aaryanshukla committed:
Update libc/test/src/math/CMakeLists.txt

Co-authored-by: OverMighty <its.overmighty@gmail.com>

aaryanshukla committed:
fixed errno.errno in smoke Cmake

overmighty reviewed (approved):
LGTM besides a last issue: FMA smoke tests having unused dependencies. You should look at the dependencies of the respective `f16*` functions and copy them unless you find that the dependency list is wrong.

aaryanshukla committed:
Update libc/test/src/math/smoke/CMakeLists.txt

Co-authored-by: OverMighty <its.overmighty@gmail.com>

aaryanshukla committed:
Update libc/test/src/math/smoke/CMakeLists.txt

Co-authored-by: OverMighty <its.overmighty@gmail.com>

aaryanshukla commented:
@overmighty I am removing dsubl for now in another patch.

mikhailramalho commented:
@aaryanshukla this is causing an issue in the rv32 buildbot: https://lab.llvm.org/staging/#/builders/196/builds/136

aaryanshukla commented:
https://github.com/llvm/llvm-project/pull/101719/files
